{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/admin/repos/sbol_llm\n"
     ]
    }
   ],
   "source": [
    "%cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from openai import OpenAI\n",
    "OPEN_AI_API_KEY = \"\"\n",
    "client = OpenAI(api_key=OPEN_AI_API_KEY)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Upload knowledge base to OpenAI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "assistant_knowledge_base_files = {'docs/fair_publication.pdf': '', \n",
    "                                  'docs/miriam_publication.pdf': ''}\n",
    "\n",
    "for file_path in assistant_knowledge_base_files.keys():\n",
    "    with open(file_path, 'rb') as file:\n",
    "        response = client.files.create(file=file, purpose='assistants')\n",
    "        assistant_knowledge_base_files[file_path] = response.id\n",
    "assistant_knowledge_base_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define System Prompt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SYSTEM_PROMPT = f\"\"\"\n",
    "    You are a genetic circuit description generator. You will be given the components of a genetic circuit, with annotations for the parts which may include, the specific gene, the part type according to standard ontology, links for annotations, short descriptions of the parts. Your job is to describe the circuit in plain English. \n",
    "    \n",
    "The description you provide must meet the following criteria:\n",
    "    \n",
    " 1. The description must be in plain English.\n",
    " 2. The description must meet FAIR data principles.\n",
    " 3. The description must provide the minimum information requested in the annotation of biochemical models (MIRIAM).\n",
    "4. The descriptions must be terse, ensuring that every work is informative.\n",
    "    \n",
    "Details for FAIR data principles and MIRIAM compliance can be obtained from the provided files. Here is a summary:\n",
    "\n",
    "MIRIAM guidelines: The Minimum Information Requested in the Annotation of Models (MIRIAM) defines standards for ensuring the accuracy, usability, and reusability of biochemical models. The key requirements are:\n",
    "\n",
    "Reference description linkage: Each part, component and the interactions between parts and components must be explicitly associated with a unique reference description that explains the biological context, structure, and expected outcomes.\n",
    "Simulation reproducibility: The description must include all relevant quantitative attributes (e.g., initial conditions, parameters, and kinetic expressions) required to simulate the circuit's behavior.\n",
    "Annotation: The model must include clear annotations linking its components to established bioinformatics resources (e.g., Gene Ontology, KEGG, UniProt).\n",
    "\n",
    "These guidelines ensure that descriptions are accurate, well-documented, and allow for reproduce-ability.\n",
    "\n",
    "FAIR data principles: The FAIR data principles define standards for ensuring that data are Findable, Accessible, Interoperable, and Reusable. The key requirements are:\n",
    "\n",
    "Findable: Data and metadata must have a globally unique and persistent identifier (e.g., DOI, accession number).\n",
    "Metadata must clearly and thoroughly describe the data and include indexed terms to facilitate searchability.\n",
    "\n",
    "Accessible: Data and metadata must be retrievable by a standard protocol (e.g., HTTP, FTP) that is open, free, and universally implementable.\n",
    "Access to metadata should persist even if the data are no longer available.\n",
    "\n",
    "Interoperable: Data and metadata must use a formal, accessible, and widely applicable language for knowledge representation (e.g., RDF, XML).\n",
    "They should include references to other relevant data and use controlled vocabularies, ontologies, or thesauri to enable integration with other datasets.\n",
    "\n",
    "Reusable: Data and metadata should include detailed provenance to trace their origins and usage.\n",
    "They must comply with domain-relevant community standards and have clear licensing for reuse.\n",
    "\n",
    "\n",
    "\n",
    "  Example inputs:\n",
    "\n",
    "  INPUT EXAMPLE #1\n",
    "\n",
    "    User description: \"This part contains a reporter gene BBa_J04450, combined with OriTRP4. Used to test plasmid mobility....\"\n",
    "    \n",
    "    - PDF of biochemical model, parameters, characteristics etc (optional)\n",
    "    - Circuit/sequence\n",
    "\n",
    "    Parts:\n",
    "        BBa_J04450\n",
    "            SO:0000804 - A region that is engineered.\n",
    "            https://synbiohub.org/public/igem/BBa_J04450/1\n",
    "            Gene: Highly engineered mutant of red fluorescent protein from Discosoma striata (coral)\n",
    "\n",
    "        OriTRP4\n",
    "            SO_0000724 - A region of a DNA molecule where transfer is initiated during the process of conjugation or mobilization.\n",
    "            https://synbiohub.org/public/igem/BBa_K1439000/1\n",
    "\n",
    "\n",
    "OUTPUT  EXAMPLE #1:\n",
    "\n",
    "This circuit contains the reporter gene, RFP (https://synbiohub.org/public/igem/BBa_J04450/1) combined with the origin of transfer for the RP4-plasmid nic region, OriTRP4 (https://synbiohub.org/public/igem/BBa_K1439000/1). \n",
    "OriTRP4 was derived from the RP4 plasmid (https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=2503).\n",
    "\n",
    "\n",
    "INPUT EXAMPLE #2:    \n",
    "\n",
    "User Description: RFP and RBS behind Lamda Prm Promoter\n",
    "\n",
    "Parts:\n",
    "    BBa_S03873\n",
    "        http://purl.obolibrary.org/obo/SO:0000804\n",
    "        Engineered Region\n",
    "        https://synbiohub.org/public/igem/BBa_S03873/1\n",
    "\n",
    "        BBa_I12006\n",
    "            Name: Prm+\n",
    "            Promoter\n",
    "            https://www.ebi.ac.uk/ols4/ontologies/so/classes?obo_id=SO%3A0000167\n",
    "            https://synbiohub.org/public/igem/BBa_I12006/1\n",
    "\n",
    "        BBa_B0034\n",
    "            https://www.ebi.ac.uk/ols4/ontologies/so/classes?obo_id=SO%3A0000139\n",
    "            Ribosomal Binding Site\n",
    "            https://synbiohub.org/public/igem/BBa_B0034/1\n",
    "        \n",
    "        BBa_E1010\n",
    "            Name: mRFP1\n",
    "            CDS\n",
    "            https://www.ebi.ac.uk/ols4/ontologies/so/classes?obo_id=SO%3A0000316\n",
    "            Highly engineered mutant of red fluorescent protein from Discosoma striata (coral)\n",
    "            https://synbiohub.org/public/igem/BBa_E1010/1\n",
    "\n",
    "OUTPUT:\n",
    "\n",
    "This circuit contains a Red fluorescent Protein (https://synbiohub.org/public/igem/BBa_E1010/1) and Ribosome Binding Site (https://synbiohub.org/public/igem/BBa_B0034/1) behind the modified Lambda Prm Promoter (https://synbiohub.org/public/igem/BBa_I12006/1).\n",
    "This modified Lambda Promoter is induced by the lamda cl repressor (https://parts.igem.org/Part:BBa_C0051) instead of repressed by it.\n",
    "When induced, the RFP is expressed. This allows the inducible Lamda Prm promoter to control the expression of an RFP.\n",
    "\n",
    "Respond only with the description and no preamble or explanation.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create agent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "my_assistant = client.beta.assistants.create(\n",
    "    instructions=SYSTEM_PROMPT,\n",
    "    name=\"Circuit Describer\",\n",
    "    tools=[{\"type\": \"code_interpreter\"}],\n",
    "    model=\"gpt-4o\",\n",
    ")\n",
    "print(my_assistant)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example function use (search NCBI for gene information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene Name: lacI\n",
      "Gene ID: 945007\n",
      "Description: DNA-binding transcriptional repressorLacI[Escherichia coli str. K-12 substr. MG1655]\n",
      "Location: NC_000913.3 (366428..367510, complement)\n",
      "Aliases: b0345, ECK0342\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/945007\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 75170321\n",
      "Description: DNA-binding transcriptional repressorLacI[Shigella dysenteriae]\n",
      "Location: \n",
      "Aliases: HUZ68_RS05005, HUZ68_04945\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/75170321\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 75202508\n",
      "Description: DNA-binding transcriptional repressorLacI[Shigella boydii]\n",
      "Location: \n",
      "Aliases: CEH00_RS02965, CEH00_02975\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/75202508\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 69758136\n",
      "Description: DNA-binding transcriptional repressorLacI[Klebsiella quasivariicola]\n",
      "Location: \n",
      "Aliases: LGM30_RS27025, LGM30_27075\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/69758136\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 60785781\n",
      "Description: DNA-binding transcriptional repressorLacI[Aeromonas allosaccharophila]\n",
      "Location: \n",
      "Aliases: I6G90_RS09205, I6G90_09205\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/60785781\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 69640750\n",
      "Description: DNA-binding transcriptional repressorLacI[Hafnia paralvei]\n",
      "Location: \n",
      "Aliases: K9N75_RS21885, K9N75_21885\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/69640750\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 57345527\n",
      "Description: DNA-binding transcriptional repressorLacI[Pantoea brenneri]\n",
      "Location: \n",
      "Aliases: HZS23_RS10170, HZS23_10175\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/57345527\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 914500\n",
      "Description: lactose operon repressor [Escherichia coli O157:H7 str. Sakai]\n",
      "Location: NC_002695.2 (423205..424296, complement)\n",
      "Aliases: ECs_0398\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/914500\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 40093722\n",
      "Description: DNA binding protein [Escherichia phage DE3]\n",
      "Location: NC_042057.1 (341..1423)\n",
      "Aliases: FDI87_gp02, ECD_10002\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/40093722\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 93777109\n",
      "Description: DNA-binding transcriptional repressorLacI[Shigella sonnei]\n",
      "Location: \n",
      "Aliases: C1P89_RS13975\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/93777109\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 93254127\n",
      "Description: DNA-binding transcriptional repressorLacI[Klebsiella quasipneumoniae]\n",
      "Location: \n",
      "Aliases: LA349_RS26240, LA349_26240\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/93254127\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 57431848\n",
      "Description: DNA-binding transcriptional repressorLacI[Raoultella planticola]\n",
      "Location: \n",
      "Aliases: MC50_RS28040, MC50_028040\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/57431848\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 86998781\n",
      "Description: DNA-binding transcriptional repressorLacI[Phytobacter diazotrophicus]\n",
      "Location: \n",
      "Aliases: QUE80_RS29260, PDTA9759_57850\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/86998781\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 86862907\n",
      "Description: DNA-binding transcriptional repressorLacI[Escherichia ruysiae]\n",
      "Location: \n",
      "Aliases: RGV86_RS17985, RGV86_17985\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/86862907\n",
      "\n",
      "Gene Name: lacI\n",
      "Gene ID: 84634240\n",
      "Description: DNA-binding transcriptional repressorLacI[Mixta calida]\n",
      "Location: \n",
      "Aliases: FS650_RS20145\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/84634240\n",
      "\n",
      "Gene Name: TFPI\n",
      "Gene ID: 7035\n",
      "Description: tissue factor pathway inhibitor [Homo sapiens(human)]\n",
      "Location: Chromosome 2, NC_000002.12 (187464230..187554435, complement)\n",
      "Aliases: EPI,LACI, TFI1, TFPI\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/7035\n",
      "\n",
      "Gene Name: Tfpi\n",
      "Gene ID: 21788\n",
      "Description: tissue factor pathway inhibitor [Mus musculus(house mouse)]\n",
      "Location: Chromosome 2, NC_000068.8 (84263199..84307119, complement)\n",
      "Aliases: A630013F22Rik, EPI,LACI\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/21788\n",
      "\n",
      "Gene Name: TFPI\n",
      "Gene ID: 100009401\n",
      "Description: tissue factor pathway inhibitor [Oryctolagus cuniculus(rabbit)]\n",
      "Location: Chromosome 7, NC_067380.1 (45699067..45765539)\n",
      "Aliases: EPI,LACI\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/100009401\n",
      "\n",
      "Gene Name: TFPI\n",
      "Gene ID: 613026\n",
      "Description: tissue factor pathway inhibitor [Macaca mulatta(Rhesus monkey)]\n",
      "Location: Chromosome 12, NC_041765.1 (74605471..74654000, complement)\n",
      "Aliases: EPI,LACI\n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/613026\n",
      "\n",
      "Gene Name: Tfpi\n",
      "Gene ID: 29436\n",
      "Description: tissue factor pathway inhibitor [Rattus norvegicus(Norway rat)]\n",
      "Location: Chromosome 3, NC_086021.1 (89939862..89989253, complement)\n",
      "Aliases: \n",
      "Link: https://www.ncbi.nlm.nih.gov/gene/29436\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import json\n",
    "import time\n",
    "from requests import RequestException\n",
    "\n",
    "def search_ncbi_gene(gene_name):\n",
    "    \"\"\"\n",
    "    Search NCBI Gene database for a given gene name.\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ncbi.nlm.nih.gov\"\n",
    "    search_url = f\"{base_url}/gene/?term={gene_name}\"\n",
    "    response = requests.get(search_url)\n",
    "\n",
    "    if response.status_code == 200:\n",
    "        return parse_ncbi_gene_html(response.text)\n",
    "    else:\n",
    "        return {\"error\": f\"Failed to access NCBI Gene. Status code: {response.status_code}\"}\n",
    "\n",
    "\n",
    "def parse_ncbi_gene_html(html_content):\n",
    "    \"\"\"\n",
    "    Parse the NCBI Gene HTML content to extract gene-related information.\n",
    "    \"\"\"\n",
    "\n",
    "    soup = BeautifulSoup(html_content, \"html.parser\")\n",
    "    results = []\n",
    "\n",
    "    table = soup.find(\"table\", class_=\"jig-ncbigrid gene-tabular-rprt\")\n",
    "    if not table:\n",
    "        return {\"error\": \"Results table not found in the HTML.\"}\n",
    "\n",
    "    rows = table.find_all(\"tr\", class_=\"rprt\")\n",
    "    for row in rows:\n",
    "        gene_data = {}\n",
    "        \n",
    "        name_cell = row.find(\"td\", class_=\"gene-name-id\")\n",
    "        if name_cell:\n",
    "            name_link = name_cell.find(\"a\", href=True)\n",
    "            if name_link:\n",
    "                gene_data[\"name\"] = name_link.get_text(strip=True)\n",
    "                gene_data[\"link\"] = f\"https://www.ncbi.nlm.nih.gov{name_link['href']}\"\n",
    "            gene_id = name_cell.find(\"span\", class_=\"gene-id\")\n",
    "            if gene_id:\n",
    "                gene_data[\"id\"] = gene_id.get_text(strip=True).replace(\"ID: \", \"\")\n",
    "        \n",
    "        # Extract the description\n",
    "        description_cell = row.find_all(\"td\")[1] if len(row.find_all(\"td\")) > 1 else None\n",
    "        if description_cell:\n",
    "            gene_data[\"description\"] = description_cell.get_text(strip=True)\n",
    "        \n",
    "        # Extract the location\n",
    "        location_cell = row.find_all(\"td\")[2] if len(row.find_all(\"td\")) > 2 else None\n",
    "        if location_cell:\n",
    "            gene_data[\"location\"] = location_cell.get_text(strip=True)\n",
    "        \n",
    "        # Extract aliases\n",
    "        aliases_cell = row.find_all(\"td\")[3] if len(row.find_all(\"td\")) > 3 else None\n",
    "        if aliases_cell:\n",
    "            gene_data[\"aliases\"] = aliases_cell.get_text(strip=True)\n",
    "        \n",
    "        if gene_data:\n",
    "            results.append(gene_data)\n",
    "\n",
    "    if results:\n",
    "        return results\n",
    "    else:\n",
    "        return {\"error\": \"No gene data found in the HTML.\"}\n",
    "\n",
    "# Example Usage\n",
    "gene_name = \"lacI\"\n",
    "gene_results = search_ncbi_gene(gene_name)\n",
    "\n",
    "if \"error\" in gene_results:\n",
    "    print(gene_results[\"error\"])\n",
    "else:\n",
    "    for gene in gene_results:\n",
    "        print(f\"Gene Name: {gene['name']}\")\n",
    "        print(f\"Gene ID: {gene.get('id', 'N/A')}\")\n",
    "        print(f\"Description: {gene.get('description', 'No description available')}\")\n",
    "        print(f\"Location: {gene.get('location', 'No location available')}\")\n",
    "        print(f\"Aliases: {gene.get('aliases', 'No aliases available')}\")\n",
    "        print(f\"Link: {gene['link']}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Configure the assistant to use the search_ncbi_gene function\n",
    "tools = [\n",
    "        {\n",
    "            \"type\": \"function\",\n",
    "            \"function\": {\n",
    "                \"name\": \"search_ncbi_gene\",\n",
    "                \"description\": \"Search for gene information and links\",\n",
    "                \"parameters\": {\n",
    "                    \"type\": \"object\",\n",
    "                    \"properties\": {\n",
    "                        \"query\": {\n",
    "                            \"type\": \"string\",\n",
    "                            \"description\": \"gene name or symbol\",\n",
    "                        },\n",
    "                    },\n",
    "                    \"required\": [\"query\"],\n",
    "                },\n",
    "            },\n",
    "        }\n",
    "    ]\n",
    "\n",
    "my_assistant = client.beta.assistants.update(\n",
    "    assistant_id=my_assistant.id,\n",
    "    tools=tools,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
